### PR TITLE
Overlay support in AGIC

### PR DESCRIPTION
**Pull request by akshaysngupta**
## Checklist
- [x] The title of the PR is clear and informative
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description

This pull request introduces overlay support in the AGIC, specifically addressing overlay changes. It also refactors the Kubenet CNI.

In Overlay cluster, AGIC will need to create an Overlay Extension Config in which it will specify the Subnet CIDR where the gateway is deployed. This config then gets processed by the AKS cluster to allow AppGW data plane to be able to reach the Pods.

### CODE REVIEW

File: helm/ingress-azure/templates/clusterrole.yaml [hunk 1]
```diff
@@ -65,4 +65,22 @@ rules:
   verbs:
     - create
     - patch
+- apiGroups:
+    - "acn.azure.com"
+  resources:
+    - "overlayextensionconfigs"
```
Comment (snehachhabria) [inline helm/ingress-azure/templates/clusterrole.yaml:71]: does the crd for this need to be installed on the cluster prior to use of this feature or install of AGIC (with this feature) ?

File: pkg/k8s/scheme.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	overlayextensionconfig_v1alpha1 "github.com/Azure/azure-container-networking/crd/overlayextensionconfig/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// NewScheme builds and returns k8s schemes used by ALB Controller.
+func NewScheme() (*runtime.Scheme, error) {
+	s := runtime.NewScheme()
+	sb := runtime.SchemeBuilder{
+		// Azure CNI CRDs
+		overlayextensionconfig_v1alpha1.AddToScheme,
```
Comment (snehachhabria) [inline pkg/k8s/scheme.go:14]: are these two crds present on the cluster by default ?

**Review by snehachhabria — APPROVED** [review]

File: pkg/cni/overlay.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+		klog.Infof("Creating overlay extension config with subnet CIDR %s", subnetCIDR)
+		err := r.client.Create(ctx, &config)
+		if err != nil {
+			return errors.Wrap(err, "failed to create overlay extension config")
+		}
+		return nil
+	}
+
+	config.Spec.ExtensionIPRange = subnetCIDR
```
Comment (robogatikov) [inline pkg/cni/overlay.go:87]: ExtensionIPRange field is going to be immutable. Will this cause issues?
https://github.com/Azure/azure-container-networking/pull/3056

File: cmd/appgw-ingress/main.go [hunk 4]
```diff
@@ -199,18 +215,11 @@ func main() {
...
-		subnetID := *(*appGw.GatewayIPConfigurations)[0].Subnet.ID
-		routeTableID := azure.RouteTableID(azure.SubscriptionID(cpConfig.SubscriptionID), azure.ResourceGroup(cpConfig.RouteTableResourceGroup), azure.ResourceName(cpConfig.RouteTableName))
-
-		err = azClient.ApplyRouteTable(subnetID, routeTableID)
-		if err != nil {
-			klog.V(3).Infof("Unable to associate Application Gateway subnet '%s' with route table '%s' due to error (this is relevant for AKS clusters using 'Kubenet' network plugin): [%+v]",
-				subnetID,
-				routeTableID,
-				err)
+	if err := cni.ReconcileCNI(context.Background(), azClient, ctrlClient, env.AGICPodNamespace, cpConfig, appGw); err != nil {
```
Comment (robogatikov) [inline cmd/appgw-ingress/main.go:218]: @nairashu was also suggesting to validate at AGIC level that VNET ID is the same for OverlayExtensionConfig ExtensionIPRange cidr and for aks cluster node ip addresses.

File: pkg/cni/overlay.go (hunk 3 above)
Comment (akshaysngupta) [reply pkg/cni/overlay.go:87]: Thanks for sharing that. What happens if user delete and creates a config with the same name but different IP? Is that an issue ?

File: cmd/appgw-ingress/main.go (hunk 4 above)
Comment (akshaysngupta) [reply cmd/appgw-ingress/main.go:218]: Ok. Do you what is the best way to get the Cluster VNET ID on the controller ?

File: pkg/k8s/scheme.go (hunk 2 above)
Comment (akshaysngupta) [reply pkg/k8s/scheme.go:14]: These two are suppose to be present only on overlay clusters.

File: pkg/cni/overlay.go (hunk 3 above)
Comment (robogatikov) [reply pkg/cni/overlay.go:87]: If user deletes OEC custom resource, clean up logic will run that will delete all the programmed Network Containers. After that OEC CR will be deleted. So, I don't think that this is an issue

File: cmd/appgw-ingress/main.go (hunk 4 above)
Comment (akshaysngupta) [reply cmd/appgw-ingress/main.go]: Will take this up as a separate pr

**Review by snehachhabria — APPROVED** [review]

**Comment by michael-rechani** [comment]
Running into the same issue here - https://github.com/Azure/application-gateway-kubernetes-ingress/issues/1524.  Happy to test AGIC with AKS and Azure CNI overlay once the update has been made.

**Pull request merged**